### PR TITLE
Revamp Library page

### PR DESCRIPTION
**Pull request by mirkobrombin**
# Preview
https://user-images.githubusercontent.com/9479903/202860074-cde369c0-86b2-430b-be61-80c50e741028.mp4

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
- [x] Local build, check the video

### CODE REVIEW

**Review by TheEvilSkeleton — CHANGES_REQUESTED** [review]
A couple of suggestions:
1. A toast that notifies the user that the entry is running
2. A toast that notifies the user when they remove an entry (and an option to undo)
3. Since we're discussing about refactoring, the library should only display a maximum of 3 entries per row

**Comment by mirkobrombin** [comment]
> A couple of suggestions:
> 
> 1. A toast that notifies the user that the entry is running
> 2. A toast that notifies the user when they remove an entry (and an option to undo)
> 3. Since we're discussing about refactoring, the library should only display a maximum of 3 entries per row
> 4. Entries should be placed at the center and not at the far left (like GNOME Software)

3xRow looks so wrong
image

**Comment by mirkobrombin** [comment]
The problem with the trash icon is that it can be intended as "completely remove this program"

**Comment by jannuary** [comment]
I don't think this is an issue within this context — meanwhile, – would be confusing.

**Comment by TheEvilSkeleton** [comment]
> > A couple of suggestions:
> > 
> > 1. A toast that notifies the user that the entry is running
> > 2. A toast that notifies the user when they remove an entry (and an option to undo)
> > 3. Since we're discussing about refactoring, the library should only display a maximum of 3 entries per row
> > 4. Entries should be placed at the center and not at the far left (like GNOME Software)
> 
> 3xRow looks so wrong image

IMO it's better to have _some_ restrictions than none. Perhaps 4 or 5 would look better? Personally, I like how it looks with 3.

File: bottles/frontend/ui/library-entry.blp [hunk 1]
```diff
@@ -69,77 +42,113 @@ template LibraryEntry : Box {
...
+            ]
+          }
+        }
+
+        [end]
+        Box {
+          Button btn_remove {
+            halign: center;
+            icon-name: "list-remove-symbolic";
+            tooltip-text: _("Remove");
```
Comment (TheEvilSkeleton) [inline bottles/frontend/ui/library-entry.blp:130]: Since @tiziodcaio and @jannuary are in favor for a trash icon, perhaps we could update the tooltip text?

```suggestion
            tooltip-text: _("Remove from library");
```

**Review by TheEvilSkeleton — CHANGES_REQUESTED** [review]

**Comment by MrQuerter** [comment]
Any thoughts about having the option to choose cover sizes?

Here's few examples that have way smaller covers than Bottles which makes it great in case someone has many entries in library:
https://www.gog.com/galaxy
https://store.steampowered.com/libraryupdate

File: bottles/frontend/ui/library-entry.blp (hunk 1 above)
Comment (orowith2os) [reply bottles/frontend/ui/library-entry.blp:130]: small grammar change: Remove from Library

**Comment by TheEvilSkeleton** [comment]
> Any thoughts about having the option to choose cover sizes?
> 
> Here's few examples that have way smaller covers than Bottles which makes it great in case someone has many entries in library: https://www.gog.com/galaxy https://store.steampowered.com/libraryupdate

I'm not against it, but this is out of scope of this MR.

**Comment by orowith2os** [comment]
@TheEvilSkeleton actually, I'd argue that that's *perfect* for this MR, since it *is* titled "Library UI refactoring"

**Comment by TheEvilSkeleton** [comment]
Refactoring implies that there's no altering functionality... that being said we _are_ altering functionality so the title is wrong. :sweat_smile:

**Comment by mirkobrombin** [comment]
My intention was not altering the functionality. If you look at my initial commits, there was only UI changes. I think we should merge this and open another PR since I’m not going to take the functionality task

**Comment by mirkobrombin** [comment]
> > > A couple of suggestions:
> > > 
> > > 1. A toast that notifies the user that the entry is running
> > > 2. A toast that notifies the user when they remove an entry (and an option to undo)
> > > 3. Since we're discussing about refactoring, the library should only display a maximum of 3 entries per row
> > > 4. Entries should be placed at the center and not at the far left (like GNOME Software)
> > 
> > 
> > 3xRow looks so wrong
> 
> IMO it's better to have _some_ restrictions than none. Perhaps 4 or 5 would look better? Personally, I like how it looks with 3.

When all my programs are installed (16) I have to scroll several times limiting the screen. I agree with centring them but I would not limit the number per line.

**Comment by TheEvilSkeleton** [comment]
What about a larger number like 6? I use Bottles while maximized and having entries all the way to the far left on a 32" 1440p monitor is painful.

**Comment by mirkobrombin** [comment]
Since the time will come when we will have to implement a search bar, why not leave it like this and add the search as a task in another MR? I'm generally against making the FlexBox fixed but this is just my opinion

**Comment by TheEvilSkeleton** [comment]
Sure.

**Comment by TheEvilSkeleton** [comment]
Can you add an Undo functionality? If that's too difficult, then a prompt that asks the user if they really want to remove from library.

image

**Review by TheEvilSkeleton — CHANGES_REQUESTED** [review]
Can you add an Undo functionality? If that's too difficult, then a prompt that asks the user if they really want to remove from library.

image

**Comment by mirkobrombin** [comment]
Can’t work on this soon, at least before the release day. Btw a undo is not easily implementable since the library is a yaml file so we need to backup in memory and make a queue management to handle multiple deletions.

**Comment by TheEvilSkeleton** [comment]
I figured. What about a prompt that asks the user if they are sure they want to remove? If you don't have the time to work on it then I can commit.

**Comment by jannuary** [comment]
You would hide the entry from the library screen, then only actually remove it after getting a specific response from the toast.

**Comment by mirkobrombin** [comment]
> You would hide the entry from the library screen, then only actually remove it after getting a specific response from the toast.

But we still need a queue management for multiple deletions, since if you queue 1 deletion, that one will be hidden but not removed from the yaml file, then you ask for another deletion and lose the prior one since you should always be aware of the yaml changes before editing

File: bottles/frontend/ui/style.css [hunk 2]
```diff
@@ -164,4 +163,4 @@ entry.heading_1 {
 
 .bottom-bar > box > button:last-child {
   border-radius: 0 0 12px 0;
-}
\ No newline at end of file
+}
```
Comment (TheEvilSkeleton) [inline bottles/frontend/ui/style.css:166]:
```suggestion
}

```

File: com.usebottles.bottles.dev.yml [hunk 3]
```diff
@@ -19,7 +19,6 @@ finish-args:
   - --system-talk-name=org.freedesktop.UDisks2
   - --env=LD_LIBRARY_PATH=/app/lib:/app/lib32
   - --env=PATH=/app/bin:/app/utils/bin:/usr/bin:/usr/lib/extensions/vulkan/MangoHud/bin/:/usr/bin:/usr/lib/extensions/vulkan/OBSVkCapture/bin/
-  - --env=GST_PLUGIN_SYSTEM_PATH=/app/lib/gstreamer-1.0:/usr/lib/x86_64-linux-gnu/gstreamer-1.0:/app/lib32/gstreamer-1.0:/usr/lib/i386-linux-gnu/gstreamer-1.0
```
Comment (TheEvilSkeleton) [inline com.usebottles.bottles.dev.yml:22]: This is out of scope of this MR I believe.

File: com.usebottles.bottles.dev.yml [hunk 4]
```diff
@@ -404,4 +388,4 @@ modules:
       - -Ddevel=true
     sources:
       - type: dir
-        path: .
+        path: .
```
Comment (TheEvilSkeleton) [inline com.usebottles.bottles.dev.yml:391]:
```suggestion
        path: .

```

File: com.usebottles.bottles.dev.yml [hunk 5]
```diff
@@ -245,21 +244,6 @@ modules:
...
-        url: https://mesa.freedesktop.org/archive/glu/glu-9.0.2.tar.xz
-        sha256: 6e7280ff585c6a1d9dfcdf2fca489251634b3377bfc33c29e4002466a38d02d4
-
-  - name: mesa-glu-32bit
-    build-options:
-      arch:
-        x86_64: *compat_i386_opts
-    config-opts: *mesa_glu_opts
-    sources: *mesa_glu_sources
-
```
Comment (TheEvilSkeleton) [inline com.usebottles.bottles.dev.yml:262]: This is out of scope of this MR as well.

File: com.usebottles.bottles.dev.yml [hunk 6]
```diff
@@ -394,8 +378,8 @@ modules:
       - cp -a * /app/etc/runtime/
     sources:
       - type: archive
-        url: https://github.com/bottlesdevs/runtime/releases/download/0.5/runtime.tar.gz
-        sha256: 00f7e57f0b54057343cfe49de52690c7b3024771478fc75cfe55c2154e13834f
+        url: https://github.com/bottlesdevs/runtime/releases/download/0.6/runtime.tar.gz
+        sha256: f2b47012cde0a7e9c65ca4d793d19a2f9d3fe68b76dd457839e62ec34f214528
```
Comment (TheEvilSkeleton) [inline com.usebottles.bottles.dev.yml:382]: This is out of scope of this MR as well.

File: bottles/frontend/windows/main_window.py [hunk 7]
```diff
@@ -293,9 +293,20 @@ def lock_ui(self, status: bool = True):
         for w in widgets:
             w.set_visible(not status)
 
-    def show_toast(self, message, timeout=3):
+    def show_toast(self, message, timeout=3, action_label=None, action_callback=None, dismissed_callback=None) -> Adw.Toast:
```
Comment (TheEvilSkeleton) [inline bottles/frontend/windows/main_window.py:296]:
```suggestion
    def show_toast(self,
                   message,
                   timeout=3,
                   action_label=None,
                   action_callback=None,
                   dismissed_callback=None
                   ) -> Adw.Toast:

```

**Review by TheEvilSkeleton — CHANGES_REQUESTED** [review]
I like the UX so far. There are some suggestions.
